src/libc/memmove.src

```diff
@@ -0,0 +1,107 @@
+	assume	adl=1
+
+	section	.text
+
+	public	_memmove
+
+if PREFER_OS_LIBC
+
+_memmove := 0000A8h
+
+else
+
+if 1
+
+; Optimized for the no-overlap case
+_memmove:
+	; no-overlap | LDIR | 32F + 12R + 1
+	; src > dst  | LDIR | 32F + 12R + 1
+	; src < dst  | LDDR | 36F + 9R + 2
+	; src = dst  | LDDR | 36F + 9R + 2
+	; zero size  |      | 24F + 9R + 2
+
+	ld	iy, -1
+	add	iy, sp
+	ld	bc, (iy + 10)
+	sbc	hl, hl
+	add	hl, bc
+	jr	nc, .zero
+	ld	hl, (iy + 7)
+	ld	de, (iy + 4)
+	sbc	hl, de
+	; src <= dst
+	jr	c, .copy_backwards
+; .copy_forwards:
+	add	hl, de
+	inc	hl
+	; src > dst or no-overlap
+	ldir
+.zero:
+	ld	hl, (iy + 4)
+	ret
+.copy_backwards:
+	; move HL and DE to the end
+	add	hl, de	; HL = src - 1
+	dec	de	; DE = dst - 1
+	add	hl, bc
+	ex	de, hl
+	add	hl, bc
+	ex	de, hl
```

Comment:
You can save 1 byte here by adding waiting to add DE to HL until after you've already added BC to it, like so:
```
	ex	de, hl
	add	hl, bc
	ex	de, hl	; HL = src - dst - 1, DE = dst + size
	add	hl, de	; HL = src + size - 1
	dec	de	; DE = dst + size - 1
```